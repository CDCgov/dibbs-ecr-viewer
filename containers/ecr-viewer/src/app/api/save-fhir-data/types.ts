interface Address {
  use: "home" | "work" | "temp" | "old" | "billing" | undefined;
  type: "postal" | "physical" | "both" | undefined;
  text: string | undefined;
  line: string[] | undefined;
  city: string | undefined;
  district: string | undefined;
  state: string | undefined;
  postal_code: string | undefined;
  country: string | undefined;
  period_start: Date | undefined;
  period_end: Date | undefined;
}

interface Lab {
  uuid: string | undefined;
  test_type: string | undefined;
  test_type_code: string | undefined;
  test_type_system: string | undefined;
  test_result_qualitative: string | undefined;
  test_result_quantitative: number | undefined;
  test_result_units: string | undefined;
  test_result_code: string | undefined;
  test_result_code_display: string | undefined;
  test_result_code_system: string | undefined;
  test_result_interpretation: string | undefined;
  test_result_interpretation_code: string | undefined;
  test_result_interpretation_system: string | undefined;
  test_result_ref_range_low: string | undefined;
  test_result_ref_range_low_units: string | undefined;
  test_result_ref_range_high: string | undefined;
  test_result_ref_range_high_units: string | undefined;
  specimen_type: string | undefined;
  performing_lab: string | undefined;
  specimen_collection_date: Date | undefined;
}

interface ruleSummary {
  summary: string;
}

interface RR {
  condition: string;
  rule_summaries: ruleSummary[];
}

export interface BundleExtendedMetadata {
  patient_id: string;
  person_id: string;
  gender: string | undefined;
  race: string | undefined;
  ethnicity: string | undefined;
  patient_addresses: Address[] | undefined;
  latitude: number | undefined;
  longitude: number | undefined;
  rr_id: string | undefined;
  processing_status: string | undefined;
  eicr_set_id: string | undefined;
  eicr_id: string;
  eicr_version_number: string;
  replaced_eicr_id: string | undefined;
  replaced_eicr_version: string | undefined;
  authoring_datetime: Date | undefined;
  provider_id: string | undefined;
  facility_id_number: string | undefined;
  facility_name: string | undefined;
  facility_type: string | undefined;
  encounter_type: string | undefined;
  encounter_start_date: Date | undefined;
  encounter_end_date: Date | undefined;
  reason_for_visit: string | undefined;
  active_problems: string[] | undefined;
  labs: Lab[] | undefined;
  birth_sex: string | undefined;
  gender_identity: string | undefined;
  homelessness_status: string | undefined;
  disabilities: string | undefined;
  tribal_affiliation: string | undefined;
  tribal_enrollment_status: string | undefined;
  current_job_title: string | undefined;
  current_job_industry: string | undefined;
  usual_occupation: string | undefined;
  usual_industry: string | undefined;
  preferred_language: string | undefined;
  pregnancy_status: string | undefined;
  ecr_id: string;
  last_name: string | undefined;
  first_name: string | undefined;
  birth_date: Date | undefined;
  rr: RR[] | undefined;
  report_date: Date | undefined;
}
export interface BundleMetadata {
  last_name: string;
  first_name: string;
  birth_date: string;
  data_source: string;
  eicr_set_id: string | undefined;
  eicr_version_number: string | undefined;
  rr: RR[] | undefined;
  report_date: string;
}
