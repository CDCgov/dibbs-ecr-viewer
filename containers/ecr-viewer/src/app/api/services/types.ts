import {
  ColumnType,
  Generated,
  Insertable,
  Selectable,
  Updateable,
} from "kysely";

export interface ecr_data {
  eICR_ID: Generated<string>;
  set_id: string | undefined;
  data_source: string | undefined;
  fhir_reference_link: string | undefined;
  eicr_version_number: string | undefined;

  patient_name_first: string;
  patient_name_last: string;
  patient_birth_date: ColumnType<Date, string | undefined, never>;

  date_created: ColumnType<Date, string | undefined, never>;
  report_date: ColumnType<Date, string>;
}

export interface ecr_rr_conditions {
  uuid: Generated<string>;
  eICR_ID: string;
  condition: string;
}

export interface ecr_rr_rule_summaries {
  uuid: Generated<string>;
  ecr_rr_conditions_id: string;
  rule_summary: string;
}

export type ECR = Selectable<ecr_data>;
export type NewECR = Insertable<ecr_data>;
export type ECRUpdate = Updateable<ecr_data>;

export type ECRConditions = Selectable<ecr_rr_conditions>;
export type NewECRConditions = Insertable<ecr_rr_conditions>;
export type ECRConditionsUpdate = Updateable<ecr_rr_conditions>;

export type ECRRuleSummaries = Selectable<ecr_rr_rule_summaries>;
export type NewECRRuleSummaries = Insertable<ecr_rr_rule_summaries>;
export type ECRRuleSummariesUpdate = Updateable<ecr_rr_rule_summaries>;

export interface Core {
  ecr_data: ecr_data;
  ecr_rr_conditions: ecr_rr_conditions;
  ecr_rr_rule_summaries: ecr_rr_rule_summaries;
}
