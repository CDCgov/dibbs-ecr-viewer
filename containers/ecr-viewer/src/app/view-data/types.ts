import { Accordion } from "@trussworks/react-uswds";

export type AccordionItem = React.ComponentProps<typeof Accordion>["items"][0];
