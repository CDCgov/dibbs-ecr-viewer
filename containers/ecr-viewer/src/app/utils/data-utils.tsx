import React, { Fragment } from "react";
import { DisplayDataProps } from "@/app/view-data/components/DataDisplay";
import sanitizeHtml from "sanitize-html";
import parse from "html-react-parser";
import { CodeableConcept } from "fhir/r4";

export interface PathMappings {
  [key: string]: string;
}

export interface CompleteData {
  availableData: DisplayDataProps[];
  unavailableData: DisplayDataProps[];
}

export type RenderableNode = string | React.JSX.Element;

export const noData = (
  <span className="no-data text-italic text-base">No data</span>
);

/**
 * Evaluates the provided display data to determine availability.
 * @param data - An array of display data items to be evaluated.
 * @returns - An object containing arrays of available and unavailable display data items.
 */
export const evaluateData = (data: DisplayDataProps[]): CompleteData => {
  let availableData: DisplayDataProps[] = [];
  let unavailableData: DisplayDataProps[] = [];
  data.forEach((item) => {
    if (!isDataAvailable(item)) {
      unavailableData.push(item);
    } else {
      availableData.push(item);
    }
  });
  return { availableData: availableData, unavailableData: unavailableData };
};

/**
 * Checks if data is available based on DisplayDataProps value. Also filters out terms that indicate info is unavailable.
 * @param item - The DisplayDataProps object to check for availability.
 * @returns - Returns true if data is available, false otherwise.
 */
export const isDataAvailable = (item: DisplayDataProps): Boolean => {
  if (
    !item.value ||
    (Array.isArray(item.value) && item.value.length === 0) ||
    item.value === noData
  )
    return false;
  const unavailableTerms = [
    "Not on file",
    "Not on file documented in this encounter",
    "Unknown",
    "Unknown if ever smoked",
    "Tobacco smoking consumption unknown",
    "Do not know",
    "No history of present illness information available",
  ];
  const valStr = removeHtmlElements(`${item.value}`).trim();
  return !unavailableTerms.some((t) => t === valStr);
};

/**
 * Parses and sanitizes the html while also mapping common hl7v3 tags to html.
 * @param val - The string of content to parse.
 * @returns - Returns sanitized and mapped content.
 */
export const safeParse = (val: string): RenderableNode => {
  const parsed = parse(
    sanitizeHtml(val, {
      transformTags: {
        paragraph: "p",
        list: "ul",
        item: "li",
        content: "span",
      },
    }),
  );

  return Array.isArray(parsed) ? arrayToElement(parsed) : parsed;
};

/**
 * Removes HTML tags from a given string.
 * @param element - The input string containing HTML elements.
 * @returns - A string with all HTML tags removed.
 */
export const removeHtmlElements = (element: string): string => {
  return sanitizeHtml(element, {
    allowedTags: [],
    allowedAttributes: {},
  });
};

/**
 * Collapses an array of values into one element.
 * @param vals - An array of strings and elments.
 * @returns - One string or element.
 */
export const arrayToElement = (vals: RenderableNode[]) => {
  // Filter out empty nodes.
  const trimmed = vals.map(trimEmptyElements).filter(Boolean);

  // An empty array is returned for an empty input
  if (trimmed.length === 0) {
    return "";
  } else if (trimmed.length === 1) {
    return vals[0];
  }

  // Wrap the items in a fragment and make sure they all have keys.
  // It reduces the cases to handle elsewhere in the logic and avoids
  // duplicate key warnings.
  return (
    <>
      {trimmed.map((item, ind) => {
        // Make sure items always have a key
        const key = `el-${ind}`;
        return typeof item !== "object" ? (
          <Fragment key={key}>{item}</Fragment>
        ) : (
          { ...item, key }
        );
      })}
    </>
  );
};

/**
 * Return "" if an element is empty, otherwise return the element.
 * <br /> are returned as is as it shouldn't have content in it.
 * @param val - A RenderableNode.
 * @returns - A string or element.
 */
const trimEmptyElements = (val: RenderableNode) => {
  if (typeof val === "string") {
    return val.trim();
  }

  // allowed to be empty - self-closing
  if (val.type === "br") return val;

  // got children? go ahead
  return val.props.children ? val : "";
};

/**
 * Return an iterator over values from start to end by step.
 * @param start - the starting index if `end` included, or stopping index if the only param (inclusive)
 * @param end - the stopping index (exclusive)
 * @param step - the number to count by (optional)
 * @returns - An iterator of indexes.
 */
export const range = (start: number, end?: number, step: number = 1) => {
  let output = [];

  if (typeof end === "undefined") {
    end = start;
    start = 0;
  }

  for (let i = start; i < end; i += step) {
    output.push(i);
  }

  return output;
};

/**
 * Attempts to return a human-readable display value for a CodeableConcept. It will return the first
 * available value in the following order:
 * 1) `undefined` if the `CodeableConcept` is falsy
 * 2) `CodeableConcept.text`
 * 3) value of the first `coding` with a `display` value
 * 4) `code` and `system` values of the first `coding` with a `code` and `system values.
 * 6) `undefined`
 * @param codeableConcept - The CodeableConcept to get the display value from.
 * @returns - The human-readable display value of the CodeableConcept.
 */
export const getHumanReadableCodeableConcept = (
  codeableConcept: CodeableConcept | undefined,
) => {
  if (!codeableConcept) {
    return undefined;
  }

  const { coding, text } = codeableConcept;

  if (text) {
    return text;
  }

  const firstCodingWithDisplay = coding?.find((c) => c.display);
  if (firstCodingWithDisplay?.display) {
    return firstCodingWithDisplay.display;
  }

  const firstCodingWithCode = coding?.find((c) => c.code && c.system);
  if (firstCodingWithCode?.code && firstCodingWithCode?.system) {
    return `${firstCodingWithCode.code} (${firstCodingWithCode.system})`;
  }

  return undefined;
};
