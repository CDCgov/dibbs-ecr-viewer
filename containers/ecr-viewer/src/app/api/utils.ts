import * as fs from "fs";
import * as path from "path";
import yaml from "js-yaml";
import { PathMappings } from "@/app/utils/data-utils";

export const S3_SOURCE = "s3";
export const AZURE_SOURCE = "azure";

/**
 * Loads the YAML configuration for path mappings from a predefined file location.
 * @returns An object representing the path mappings defined in the YAML configuration file.
 */
export function loadYamlConfig(): PathMappings {
  const filePath = path.join(process.cwd(), "src/app/api/fhirPath.yaml");
  const fileContents = fs.readFileSync(filePath, "utf8");
  return <PathMappings>yaml.load(fileContents);
}

/**
 * Converts stream data to json data
 * @param stream - The input stream that provides JSON data in chunks. The stream
 *   should implement the async iterable protocol to allow for-await-of
 *   iteration over its data chunks.
 * @returns A promise that resolves to the JSON-parsed object from the accumulated
 *  stream data. The specific structure of this object depends on the JSON
 *  content of the stream.
 */
export async function streamToJson(stream: NodeJS.ReadableStream | undefined) {
  if (!stream) throw new Error("Stream is undefined");

  let rawData = "";
  for await (const chunk of stream) {
    rawData += chunk;
  }
  return JSON.parse(rawData);
}
