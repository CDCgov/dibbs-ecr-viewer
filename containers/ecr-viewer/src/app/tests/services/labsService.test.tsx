import { render, screen } from "@testing-library/react";
import { Bundle, DiagnosticReport, Observation, Organization } from "fhir/r4";

import { loadYamlConfig } from "@/app/api/utils";
import {
  getLabJsonObject,
  getObservations,
  checkAbnormalTag,
  searchResultRecord,
  returnFieldValueFromLabHtmlString,
  evaluateOrganismsReportData,
  evaluateDiagnosticReportData,
  evaluateObservationTable,
  isLabReportElementDataList,
  evaluateLabOrganizationData,
  ResultObject,
  combineOrgAndReportData,
  evaluateLabInfoData,
  findIdenticalOrg,
  returnAnalysisTime,
  LabReportElementData,
} from "@/app/services/labsService";
import BundleLab from "@/app/tests/assets/BundleLab.json";
import BundleLabInvalidResultsDiv from "@/app/tests/assets/BundleLabInvalidResultsDiv.json";
import BundleLabNoLabIds from "@/app/tests/assets/BundleLabNoLabIds.json";
import { noData } from "@/app/utils/data-utils";
import { evaluate } from "@/app/utils/evaluate";
import { DisplayDataProps } from "@/app/view-data/components/DataDisplay";
import { AccordionItem } from "@/app/view-data/types";

const mappings = loadYamlConfig();

const pathLabReportNormal =
  "Bundle.entry.resource.where(resourceType = 'DiagnosticReport').where(id = 'c090d379-9aea-f26e-4ddc-378223841e3b')";
const labReportNormal = evaluate(BundleLab, pathLabReportNormal)[0];
const labReportNormalJsonObject = {
  resultId: "Result.1.2.3.4.5",
  resultName: "Stool Pathogens, NAAT, 12 to 25 Targets",
  tables: [
    [
      {
        Component: {
          value: "Campylobacter, NAAT",
          metadata: {
            "data-id": "Result.1.2.3.4.5.Comp1Name",
          },
        },
        Value: { value: "Not Detected", metadata: {} },
        "Ref Range": { value: "Not Detected", metadata: {} },
        "Test Method": {
          value: <p>LAB DEVICE: BIOFIRE速 FILMARRAY速 2.0 SYSTEM</p>,
          metadata: {},
        },
        "Analysis Time": {
          value: (
            <p>
              <span>
                <i>09/28/2000 1:59:00 PM PDT</i>
              </span>
            </p>
          ),
          metadata: {},
        },
        "Performed At": {
          value: "Gungan City Hospital",
          metadata: {},
        },
        "Pathologist Signature": {
          value: "",
          metadata: {
            "data-id": "Result.1.2.3.4.5.Comp1Signature",
          },
        },
      },
      {
        Component: {
          value: "Plesiomonas shigelloides, NAAT",
          metadata: {
            "data-id": "Result.1.2.3.4.5.Comp2Name",
          },
        },
        Value: { value: "Not Detected", metadata: {} },
        "Ref Range": { value: "Not Detected", metadata: {} },
        "Test Method": {
          value: <p>LAB DEVICE: BIOFIRE速 FILMARRAY速 2.0 SYSTEM</p>,
          metadata: {},
        },
        "Analysis Time": {
          value: (
            <p>
              <span>09/28/2000 1:59:00 PM PDT</span>
            </p>
          ),
          metadata: {},
        },
        "Performed At": {
          value: "Gungan City Hospital",
          metadata: {},
        },
        "Pathologist Signature": {
          value: "",
          metadata: {
            "data-id": "Result.1.2.3.4.5.Comp2Signature",
          },
        },
      },
    ],
    [
      {
        "Specimen (Source)": {
          value: "Stool",
          metadata: {
            "data-id": "Result.1.2.3.4.5.Specimen",
          },
        },
        "Anatomical Location / Laterality": {
          value: "STOOL SPECIMEN / Unknown",
          metadata: {},
        },
        "Collection Method / Volume": { value: "", metadata: {} },
        "Collection Time": {
          value: "09/28/2000 4:51\u00A0PM\u00A0EDT",
          metadata: {},
        },
        "Received Time": {
          value: "09/28/2000 4:51\u00A0PM\u00A0EDT",
          metadata: {},
        },
      },
    ],
    [
      {
        "Authorizing Provider": { value: "Darth Test MD", metadata: {} },
        "Result Type": {
          value: "MICROBIOLOGY - GENERAL ORDERABLES",
          metadata: {},
        },
      },
    ],
    [
      {
        "Performing Organization": {
          value: <p>Gungan City Hospital</p>,
          metadata: {
            "data-id": "Result.1.2.3.4.5.PerformingLab",
          },
        },
        Address: { value: <p>500000 S. Jar Jar Binks Street</p>, metadata: {} },
        "City/State/ZIP Code": {
          value: <p>Gungan City, CA 00123</p>,
          metadata: {},
        },
        "Phone Number": { value: <p>555-555-5555</p>, metadata: {} },
      },
    ],
  ],
};

const pathLabReportAbnormal =
  "Bundle.entry.resource.where(resourceType = 'DiagnosticReport').where(id = '68477c03-5689-f9e5-c267-a3c7bdff6fe0')";
const labReportAbnormal = evaluate(BundleLab, pathLabReportAbnormal)[0];
const labReportAbnormalJsonObject = getLabJsonObject(
  labReportAbnormal,
  BundleLab as unknown as Bundle,
  mappings,
);

const pathLabOrganismsTableAndNarr =
  "Bundle.entry.resource.where(resourceType = 'DiagnosticReport').where(id = 'b0f590a6-4bf5-7add-9716-2bd3ba6defb2')";
const labOrganismsTableAndNarr = evaluate(
  BundleLab,
  pathLabOrganismsTableAndNarr,
)[0];

describe("LabsService tests", () => {
  describe("Labs Utils", () => {
    describe("getObservations", () => {
      it("extracts an array of observation resources", () => {
        const result = getObservations(
          {
            result: [
              {
                reference: "Observation/1c0f3367-0588-c90e-fed0-0d8c15c5ac1b",
              },
            ],
            resourceType: "DiagnosticReport",
            code: {},
            status: "entered-in-error",
          },
          BundleLab as unknown as Bundle,
          mappings,
        );

        const expectedObservationPath =
          "Bundle.entry.resource.where(resourceType = 'Observation').where(id = '1c0f3367-0588-c90e-fed0-0d8c15c5ac1b')";
        const expectedResult = evaluate(
          BundleLab,
          expectedObservationPath,
        ) as unknown[] as Observation[];
        expect(result.toString()).toBe(expectedResult.toString());
      });

      it("returns an empty array of observation resources if none are found", () => {
        const result = getObservations(
          {
            result: [
              {
                reference: "Observation/invalid-observation-id",
              },
            ],
            resourceType: "DiagnosticReport",
            code: {},
            status: "final",
          },
          BundleLab as unknown as Bundle,
          mappings,
        );
        expect(result).toStrictEqual([]);
      });
    });

    describe("getLabJsonObject", () => {
      it("returns correct Json Object for table with data-id", () => {
        const expectedResult = labReportNormalJsonObject;

        const result = getLabJsonObject(
          labReportNormal,
          BundleLab as unknown as Bundle,
          mappings,
        );

        expect(result).toEqual(expectedResult);
      });

      it("returns correct Json Object for table without data-id", () => {
        const labReportWithoutIds: DiagnosticReport = evaluate(
          BundleLabNoLabIds,
          "Bundle.entry.resource.where(resourceType = 'DiagnosticReport').where(id = '97d3b36a-f833-2f3c-b456-abeb1fd342e4')",
        )[0];
        const labReportJsonObjectWithoutId = {
          resultId: undefined,
          resultName: "",
          tables: [
            [
              {
                "Lab Test Name": {
                  metadata: {},
                  value: <span>SARS-CoV-2, NAA CL</span>,
                },
                "Lab Test Result Date": {
                  metadata: {},
                  value: <span>2000-02-04T21:02:00.000Z</span>,
                },
                "Lab Test Result Value": {
                  metadata: {},
                  value: <span>POS</span>,
                },
              },
              {
                "Lab Test Name": {
                  metadata: {},
                  value: <span>Symptomatic as defined by CDC?</span>,
                },
                "Lab Test Result Date": {
                  metadata: {},
                  value: <span>2000-02-04T21:02:00.000Z</span>,
                },
                "Lab Test Result Value": {
                  metadata: {},
                  value: <span>YES</span>,
                },
              },
            ],
          ],
        };

        const result = getLabJsonObject(
          labReportWithoutIds,
          BundleLabNoLabIds as Bundle,
          mappings,
        );

        expect(result).toStrictEqual(labReportJsonObjectWithoutId);
      });

      it("returns empty object if lab results html contains no tables", () => {
        const result = getLabJsonObject(
          labReportNormal,
          BundleLabInvalidResultsDiv as unknown as Bundle,
          mappings,
        );

        expect(result).toStrictEqual({});
      });
    });

    describe("checkAbnormalTag", () => {
      it("should return true if lab report has abnormal tag", () => {
        const expectedResult = true;
        const result = checkAbnormalTag(labReportAbnormalJsonObject);

        expect(result).toStrictEqual(expectedResult);
      });

      it("should return false if lab report does not have abnormal tag", () => {
        const expectedResult = false;
        const result = checkAbnormalTag(labReportNormalJsonObject);

        expect(result).toStrictEqual(expectedResult);
      });
    });

    describe("searchResultRecord", () => {
      const labHTMLJson = labReportNormalJsonObject.tables;

      it("extracts string of all results of a search for specified lab report", () => {
        const searchKey = "Collection Time";
        const expectedResult = "09/28/2000 4:51\u00A0PM\u00A0EDT";

        const result = searchResultRecord(labHTMLJson, searchKey);

        expect(result).toStrictEqual(expectedResult);
      });

      it("returns an empty string of results if none are found for search key", () => {
        const invalidSearchKey = "foobar";
        const expectedResult = "";

        const result = searchResultRecord(labHTMLJson, invalidSearchKey);

        expect(result).toEqual(expectedResult);
      });
    });

    describe("returnAnalysisTime", () => {
      it("extracts and formats correct field value from within a lab report", () => {
        const fieldName = "Analysis Time";

        const result = returnAnalysisTime(labReportNormalJsonObject, fieldName);

        expect(result).toEqual("09/28/2000 4:59\u00A0PM\u00A0EDT");
      });

      it("extracts returns noData if unavailable", () => {
        const fieldName = "Analysis Time";

        const result = returnAnalysisTime({}, fieldName);

        expect(result).toEqual(noData);
      });

      it("Concats date times if multiple passed", () => {
        const fieldName = "Analysis Time";

        const result = returnAnalysisTime(
          {
            resultId: "Result.1.2.3.4.5",
            resultName: "Stool Pathogens, NAAT, 12 to 25 Targets",
            tables: [
              [
                {
                  "Analysis Time": {
                    value: (
                      <p>
                        <span>
                          <i>09/28/2000 1:59:00 PM PDT</i>
                        </span>
                        <span>
                          <i>09/28/2000 2:59:00 PM PDT</i>
                        </span>
                      </p>
                    ),
                    metadata: {},
                  },
                },
              ],
            ],
          },
          fieldName,
        );

        expect(result).toEqual(
          "09/28/2000 4:59\u00A0PM\u00A0EDT, 09/28/2000 5:59\u00A0PM\u00A0EDT",
        );
      });

      it("Returns concated date times if deeply nested", () => {
        const fieldName = "Analysis Time";

        const result = returnAnalysisTime(
          {
            resultId: "Result.1.2.3.4.5",
            resultName: "Stool Pathogens, NAAT, 12 to 25 Targets",
            tables: [
              [
                {
                  "Analysis Time": {
                    value: (
                      <p>
                        <span>
                          <i>09/28/2000 1:59:00 PM PDT</i>
                          <i>09/28/2000 2:59:00 PM PDT</i>
                        </span>
                      </p>
                    ),
                    metadata: {},
                  },
                },
              ],
            ],
          },
          fieldName,
        );

        expect(result).toEqual(
          "09/28/2000 4:59\u00A0PM\u00A0EDT, 09/28/2000 5:59\u00A0PM\u00A0EDT",
        );
      });

      it("Returns noData if emptiness is nested", () => {
        const fieldName = "Analysis Time";

        const result = returnAnalysisTime(
          {
            resultId: "Result.1.2.3.4.5",
            resultName: "Stool Pathogens, NAAT, 12 to 25 Targets",
            tables: [
              [
                {
                  "Analysis Time": {
                    value: (
                      <p>
                        <span>
                          <i></i>
                        </span>
                      </p>
                    ),
                    metadata: {},
                  },
                },
              ],
            ],
          },
          fieldName,
        );

        expect(result).toEqual(noData);
      });
    });

    describe("returnFieldValueFromLabHtmlString", () => {
      it("extracts correct field value from within a lab report", () => {
        const fieldName = "Analysis Time";

        const result = returnFieldValueFromLabHtmlString(
          labReportNormalJsonObject,
          fieldName,
        );

        expect(result).toMatchSnapshot();
      });

      it("returns NoData if none are found for field name", () => {
        const invalidFieldName = "foobar";
        const expectedNoData = (
          <span className="no-data text-italic text-base">No data</span>
        );

        const result = returnFieldValueFromLabHtmlString(
          labReportNormalJsonObject,
          invalidFieldName,
        );

        expect(result).toStrictEqual(expectedNoData);
      });
    });

    describe("evaluateOrganismsReportData", () => {
      it("should return the correct organisms table when the data exists for a lab report", () => {
        const result = evaluateOrganismsReportData(
          labOrganismsTableAndNarr,
          BundleLab as unknown as Bundle,
          mappings,
        )!;
        render(result);

        expect(
          screen.getByText("Avycaz (Ceftazidime/Avibactam)"),
        ).toBeInTheDocument();
        expect(screen.getByText("0.25: Susceptible")).toBeInTheDocument();
        expect(screen.getAllByText("MIC")).toHaveLength(3);
      });
      it("should return undefined if lab organisms data does not exist for a lab report", () => {
        const result = evaluateOrganismsReportData(
          labReportNormal,
          BundleLab as unknown as Bundle,
          mappings,
        );

        expect(result).toBeUndefined();
      });
    });
  });

  describe("Evaluate Diagnostic Report", () => {
    it("should evaluate diagnostic report results", () => {
      const report = evaluate(BundleLab, mappings.diagnosticReports)[0];
      const actual = evaluateDiagnosticReportData(
        report,
        BundleLab as unknown as Bundle,
        mappings,
      );

      render(actual);

      expect(screen.getByText("Campylobacter, NAAT")).toBeInTheDocument();
      expect(screen.getAllByText("Not Detected")).not.toBeEmpty();
    });
    it("the table should not appear when there are no results", () => {
      const diagnosticReport: DiagnosticReport = {
        resourceType: "DiagnosticReport",
        code: {
          coding: [
            {
              display: "Drugs Of Abuse Comprehensive Screen, Ur",
            },
          ],
        },
        status: "final",
      };
      const actual = evaluateObservationTable(
        diagnosticReport,
        null as unknown as Bundle,
        mappings,
        [],
      );
      expect(actual).toBeUndefined();
    });
    it("should evaluate test method results", () => {
      const report = evaluate(BundleLab, mappings.diagnosticReports)[0];
      const actual = evaluateDiagnosticReportData(
        report,
        BundleLab as unknown as Bundle,
        mappings,
      );

      render(actual);

      expect(
        screen.getAllByText("LAB DEVICE: BIOFIRE速 FILMARRAY速 2.0 SYSTEM"),
      ).not.toBeEmpty();
    });
    it("should display comment", () => {
      const report = evaluate(BundleLab, mappings.diagnosticReports)[2];
      const actual = evaluateDiagnosticReportData(
        report,
        BundleLab as unknown as Bundle,
        mappings,
      );
      render(actual!);

      expect(screen.getByText("View comment")).toBeInTheDocument();
    });
  });

  describe("Evaluate Organization with ID", () => {
    it("should return a matching org", () => {
      const result = evaluateLabOrganizationData(
        "14394818-a1e9-4882-ca8b-FAKE793bb5cc",
        BundleLab as unknown as Bundle,
        mappings,
        0,
      );
      expect(result[0].value).toEqual("Tatooine Hospital");
    });
    it("should combine the data into new format", () => {
      const testResultObject: ResultObject = {
        "Organization/22c6cdd0-bde1-e220-9ba4-2c2802f795ad": [
          {} as AccordionItem,
        ],
      };
      const result = combineOrgAndReportData(
        testResultObject,
        BundleLab as unknown as Bundle,
        mappings,
      );
      expect(result[0].organizationDisplayDataProps).toBeArray();
    });
  });

  describe("Evaluate the lab info section", () => {
    it("should return a list of LabReportElementData if the lab results in the HTML table have ID's", () => {
      const result = evaluateLabInfoData(
        BundleLab as unknown as Bundle,
        evaluate(BundleLab, mappings.diagnosticReports),
        mappings,
      );
      expect(result[0]).toHaveProperty("diagnosticReportDataItems");
      expect(result[0]).toHaveProperty("organizationDisplayDataProps");
    });

    it("should return a list of DisplayDataProps if the lab results in the HTML table do not have ID's", () => {
      const result = evaluateLabInfoData(
        BundleLabNoLabIds as unknown as Bundle,
        evaluate(BundleLabNoLabIds, mappings.diagnosticReports),
        mappings,
      );
      expect(result[0]).toHaveProperty("title");
      expect(result[0]).toHaveProperty("value");
    });

    it("should properly count the number of labs", () => {
      const result = evaluateLabInfoData(
        BundleLab as Bundle,
        evaluate(BundleLab, mappings.diagnosticReports),
        mappings,
      );
      expect(isLabReportElementDataList(result)).toBeTrue();
      const props = (result[0] as LabReportElementData)
        .organizationDisplayDataProps;
      expect(props[3].title).toEqual("Number of Results");
      expect(props[3].value).toEqual(2);
    });
  });

  describe("Find Identical Org", () => {
    const orgMappings: Organization[] = [
      {
        id: "d6930155-009b-92a0-d2b9-007761c45ad2",
        name: "Coruscant Department of Public Health",
        active: true,
        address: [
          {
            use: "work",
            city: "Sacramento",
            state: "CA",
          },
        ],
        telecom: [
          {
            use: "work",
            value: "fakeemail@example.com",
            system: "email",
          },
        ],
        resourceType: "Organization",
      },
      {
        id: "f87de327-7272-42ac-012d-58904caf7ef1",
        name: "Coruscant City Department of Public Health",
        active: true,
        resourceType: "Organization",
      },
      {
        id: "21e7aca1-7a03-43dc-15e6-8f7ee24b6613",
        name: "Mos Eisley Department of Health",
        active: true,
        resourceType: "Organization",
      },
      {
        id: "d319a926-0eb3-5847-3b21-db8b778b4f07",
        name: "Naboo University Medical Center",
        address: [
          {
            use: "work",
            city: "Naboo City",
            line: ["0000 Up Ave"],
            state: "TN",
            country: "USA",
            district: "Central",
            postalCode: "00123",
          },
        ],
        telecom: [
          {
            use: "work",
            value: "+1-555-555-5555",
            system: "phone",
          },
        ],
        resourceType: "Organization",
      },
      {
        id: "22c6cdd0-bde1-e220-9ba4-2c2802f795ad",
        name: "Mos Espa Lab",
        address: [
          {
            use: "work",
            city: "Naboo City",
            line: ["0000 Up Ave"],
            state: "TN",
            country: "USA",
            district: "Central",
            postalCode: "00123",
          },
        ],
        resourceType: "Organization",
        telecom: [
          {
            value: "+1-555-555-5555",
            system: "phone",
          },
        ],
      },
      {
        id: "e3ece69c-0968-59c9-47dd-f16db731621a",
        name: "Mos Espa Lab",
        address: [
          {
            use: "work",
            city: "Naboo City",
            line: ["0000 Up Ave"],
            state: "TN",
            country: "USA",
            district: "Central",
            postalCode: "00123",
          },
        ],
        telecom: [
          {
            value: "+1-615-875-5227",
            system: "phone",
          },
        ],
        resourceType: "Organization",
      },
      {
        id: "57fcc148-b440-3a80-749b-780325e9680d",
        name: "Moderna US, Inc.",
        resourceType: "Organization",
      },
    ];

    const matchedOrg1: Organization = {
      id: "22c6cdd0-bde1-e220-9ba4-2c2802f795ad",
      name: "Mos Espa Lab",
      address: [
        {
          use: "work",
          city: "Naboo City",
          line: ["0000 Up Ave"],
          state: "TN",
          country: "USA",
          district: "Central",
          postalCode: "00123",
        },
      ],
      resourceType: "Organization",
    };

    const matchedOrg2: Organization = {
      id: "7",
      name: "Fake Lab",
      address: [
        {
          city: "North Charleston",
          line: ["11 Fake Street", "Suite 100"],
          state: "SC",
          country: "USA",
          postalCode: "29405",
        },
      ],
      resourceType: "Organization",
    };

    it("should add telecom from matching org", () => {
      expect(matchedOrg1?.telecom).not.toBeDefined();
      expect(
        findIdenticalOrg(orgMappings, matchedOrg1)?.telecom?.[0].value,
      ).toEqual("+1-615-875-5227");
    });
    it("should not add telecom because no matching org", () => {
      expect(matchedOrg2?.telecom).not.toBeDefined();
      expect(
        findIdenticalOrg(orgMappings, matchedOrg2)?.telecom?.[0].value,
      ).not.toBeDefined();
    });
  });

  describe("isLabReportElementDataList", () => {
    it("returns true when the input is a list of LabReportElementData", () => {
      const actual = isLabReportElementDataList([
        {
          diagnosticReportDataItems: [{} as AccordionItem],
          organizationId: "test-id",
          organizationDisplayDataProps: [{} as DisplayDataProps],
        },
      ]);
      expect(actual).toBe(true);
    });

    it("returns false when the input is NOT a list of LabReportElementData", () => {
      const actual = isLabReportElementDataList([
        { title: "test-title", value: "test-value" },
      ]);
      expect(actual).toBe(false);
    });
  });
});
