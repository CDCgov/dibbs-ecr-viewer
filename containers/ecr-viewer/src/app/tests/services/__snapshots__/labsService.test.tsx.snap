// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LabsService tests Labs Utils returnFieldValueFromLabHtmlString extracts correct field value from within a lab report 1`] = `
<React.Fragment>
  <p>
    <span>
      <i>
        09/28/2000 1:59:00 PM PDT
      </i>
    </span>
  </p>
  <p>
    <span>
      09/28/2000 1:59:00 PM PDT
    </span>
  </p>
</React.Fragment>
`;
